docs/v1/P4Runtime-Spec.mdk

```diff
@@ -6269,6 +6279,314 @@ field [@ProtoAny]. At the moment, there is not any mechanism to extend the
+
+* `is_default_entry`, a boolean flag which indicates whether the table entry is
+  the default entry for the table. See [Default entry](#sec-default-entry)
+  section for more information.
+
+* `metadata`, an arbitrary `bytes` value which is opaque to the
+  target. There is no requirement of where this is stored, but it must be
+  returned by the server along with the rest of the entry when the client
+  performs a read on the entry.
+
+The `priority` field must be set to a non-zero value if the match key includes a
+ternary match (&ie; in the case of PSA if the P4Info entry for the table
+indicates that one or more of its match fields has an `OPTIONAL`, `TERNARY` or
+`RANGE` match
+type) or to zero otherwise. A higher priority number indicates that the entry
+must be given higher priority when performing a table lookup. Clients must allow
+multiple entries to be added with the same priority value.  If a packet can
+match multiple entries with the same priority, it is not deterministic in the
+data plane which entry a packet will match.  If a client wishes to make the
+matching behavior deterministic, it must use different priority values for any
+pair of table entries that the same packet matches.
+
+The `match` and `priority` fields are used to uniquely identify an entry within
+a table. Therefore, these fields cannot be modified after the entry has been
+inserted and must be provided for `MODIFY` and `DELETE` updates. When deleting
+an entry, these key fields (along with `is_default_action`) are the only fields
+considered by the server. All other fields must be ignored, even if they have
+nonsensical values (such as an invalid action field). In the case of a *keyless*
+table (the table has an empty match key), the server must reject all attempts to
+`INSERT` a match entry and return an `INVALID_ARGUMENT` error.
+
+The number of match entries that a table *should* support is indicated in P4Info
+(`size` field of `Table` message). The guarantees provided to the P4Runtime
+client are the same as the ones described in the P4~16~ specification for the
+`size` property [@P4TableProperties]. In particular, some implementations may
+not be able to always accommodate an arbitrary set of entries up to the
+requested size, and other implementations may provide the P4Runtime client with
+more entries than requested. The P4Runtime server must return
+`RESOURCE_EXHAUSTED` when a table entry cannot be inserted because of a size
+limitation. It is recommended that, for the sake of portability, P4Runtime
+clients do not try to insert additional entries once the size indicated in
+P4Info has been reached.
+
+## GenericTable Categories
+
+All tables fall into certain "categories" which define what APIs are
+applicable on the table and the APIs.
+
+  * Regular tables: Read, INSERT, MODIFY, DELETE work as intended. Entries in
+    a table can be inserted, modified and deleted.
+    * `INSERT` : The union field must be set for every `INSERT` update. If
```

Comment:
It could be beneficial to allow for omitting one or all union fields and let the target provide default value(s). This could be specified by an annotation. I'm thinking about how some DASH P4 model code tables are mapped to SAI tables, where action params are mapped to SAI attributes, Some of those attributes have defaults or can be omitted and it's hard to map the P4Runtime INSERT operation to a variable-attribute SAI call.

Reply:
Added. Please check

Reply:
Added a flow chart diagram. please check

Reply:
> Added a flow chart diagram. please check
Thanks.
Is the .png editable (for posterity)?

Reply:
replacing with svg